### ISSUE TITLE
Damaged YMap keys after consecutive redo

### ISSUE THREAD

**Issue by doodlewind** [yjs/yjs#355]
**Checklist**

* [x] This is a bug report.
* [x] This report is in the correct repository.

**Describe the bug**
Similar to #343 and #317 but with simpler reproduction path (and arguably more potential impact).

**To Reproduce**
* Add a YMap with at least 2 keys.
* Update 2 of its keys (`x` and `y` in this case, simulate element dragging in real world apps) for 3 or more times.
* Undo until the YMap is deleted.
* Redo to latest state, these 2 keys are lost.

See [code sandbox](https://codesandbox.io/s/little-waterfall-q9fx5?file=/src/index.js) or following snippet:

``` js
const doc = new Y.Doc();
const yRoot = doc.getMap();
const undoMgr = new Y.UndoManager(yRoot);

const yPoint = new Y.Map();
yPoint.set("x", 0);
yPoint.set("y", 0);
yRoot.set("a", yPoint);
undoMgr.stopCapturing();

yPoint.set("x", 100);
yPoint.set("y", 100);
undoMgr.stopCapturing();

yPoint.set("x", 200);
yPoint.set("y", 200);
undoMgr.stopCapturing();

yPoint.set("x", 300);
yPoint.set("y", 300);
undoMgr.stopCapturing();

undoMgr.undo(); // x=200, y=200
undoMgr.undo(); // x=100, y=100
undoMgr.undo(); // x=0, y=0
undoMgr.undo(); // nil

undoMgr.redo(); // x=0, y=0
undoMgr.redo(); // x=100, y=100
undoMgr.redo(); // x=200, y=200
undoMgr.redo(); // expected x=300, y=300, actually nil

console.log(yRoot.toJSON());
```

**Expected behavior**
As listed in the snippet, the affected map keys shouldn't be missing.

This can be a blocker for using Yjs as "single source of truth" model layer. And although there can be some way patching the state of view layers (say DOM elements), if we use Yjs instances for serializing data to DB on Node.js, in this case user may found the elements they updated were missing :(

**Environment Information**
- Browser and Node.js, with lastest version of Yjs.

**Additional context**
The bug only happens after the paired `yMap.set` updates for 3 times or more. I can confirm that when such update happens for 2 times or less, we got similar bug until the fix in #317.

I also did some quick search in the codebase and here are something I noticed:

* Using `deleteFilter: () => false` config in UndoManager can be an overkill workaround, since it prevents the element from being deleted when undoing.
* Inspecting `doc.store.clients.get(doc.clientID)` shows the item sequence and some correlation can be found. Say for the first `Item` representing the first YMap insertion, its `redone` links exactly to the last undo that caused the problem. But the data structure still looks obscure, for the expected redo result, are there some constraints that the data structure should follow?
* There was a `@todo` comment in `followRedone` ([link](https://github.com/yjs/yjs/commit/aeb23dbaa9effbd848601dd9ba4b1526ce35088b#diff-8e03df4677fec8fdd3103498dd8f3114030b2d18bca59ac692a122795103ab82R38)) in the previous bug fix, is it related to this issue?

We use Yjs in our collaborative app and it generally works awesome, I am willing to provide help within my capacity, thanks.

**Comment by georeith** [comment]
I also noticed that the bug was still present after 3 consecutive changes https://github.com/yjs/yjs/issues/343

**Comment by doodlewind** [comment]
@georeith I think #367 could fixed them. Can you try it out locally?

**Comment by doodlewind** [comment]
With the new refactored UndoManager released, I think we can close #304 #343 #355 now.

**Comment by dmonad** [comment]
Thank you!
